docs/grammar/Solidity.g4

```diff
@@ -0,0 +1,510 @@
+		| {!$visibilitySet}? Private {$visibilitySet = true;}
+		| {!$visibilitySet}? Internal {$visibilitySet = true;}
+		| {!$constantnessSet}? Constant {$constantnessSet = true;}
+		| {!$overrideSpecifierSet}? overrideSpecifier {$overrideSpecifierSet = true;}
+		| {!$constantnessSet}? Immutable {$constantnessSet = true;}
+	)*
+	name=identifier
+	(Assign initialValue=expression)?
+	Semicolon;
+
+/**
+ * Parameter of an event.
+ */
+eventParameter: type=typeName Indexed? name=identifier?;
+/**
+ * Definition of an event. Can occur in contracts, libraries or interfaces.
+ */
+eventDefinition:
+	Event name=identifier
+	LParen (parameters+=eventParameter (Comma parameters+=eventParameter)*)? RParen
+	Anonymous?
+	Semicolon;
+
+/**
+ * Using directive to bind library functions to types.
+ * Can occur within contracts and libraries.
+ */
+usingDirective: Using userDefinedTypeName For (Mul | typeName) Semicolon;
+/**
+ * A type name can be an elementary type, a function type, a mapping type, a user-defined type
+ * (e.g. a contract or struct) or an array type.
+ */
+typeName: elementaryTypeName[true] | functionTypeName | mappingType | userDefinedTypeName | typeName LBrack expression? RBrack;
+elementaryTypeName[boolean allowAddressPayable]: Address | {$allowAddressPayable}? Address Payable | Bool | String | Bytes | SignedIntegerType | UnsignedIntegerType | FixedBytes | Fixed | Ufixed;
+functionTypeName
+locals [boolean visibilitySet = false, boolean mutabilitySet = false]
+:
+	Function LParen (arguments=parameterList)? RParen
+	(
+		{!$visibilitySet}? visibility {$visibilitySet = true;}
+		| {!$mutabilitySet}? stateMutability {$mutabilitySet = true;}
+	)*
+	(Returns LParen returnParameters=parameterList RParen)?;
+
+/**
+ * The declaration of a single variable.
+ */
+variableDeclaration: type=typeName location=dataLocation? name=identifier;
+dataLocation: Memory | Storage | Calldata;
+
+/**
+ * Complex expression.
+ * Can be an index access, an index range access, a member access, a function call (with optional function call options),
+ * a type conversion, an unary or binary expression, a comparison or assignment, a ternary expression,
+ * a new-expression (i.e. a contract creation or the allocation of a dynamic memory array),
+ * a tuple, an inline array or a primary expression (i.e. an identifier, literal or type name).
+ */
+expression:
+	expression LBrack index=expression? RBrack # IndexAccess
+	| expression LBrack start=expression? Colon end=expression? RBrack # IndexRangeAccess
+	| expression Period (identifier | Address) # MemberAccess
+	| expression LBrace (namedArgument (Comma namedArgument)*)? RBrace # FunctionCallOptions
+	| expression callArgumentList # FunctionCall
+	| Payable callArgumentList # PayableConversion
+	| Type LParen typeName RParen # MetaType
+	| (Inc | Dec | Not | BitNot | Delete | Sub) expression # UnaryPrefixOperation
+	| expression (Inc | Dec) # UnarySuffixOperation
+	|<assoc=right> expression Exp expression # ExpOperation
```

Comment:
I think this is actually not true and we should fix it in the language: https://github.com/ethereum/solidity/issues/9637